gradle-failure-reports/src/main/java/com/palantir/gradle/failurereports/CompileFailuresService.java

```diff
@@ -0,0 +1,162 @@
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.api.tasks.compile.AbstractCompile;
+
+public abstract class CompileFailuresService implements BuildService<Parameters>, AutoCloseable {
+
+    interface Parameters extends BuildServiceParameters {
+        RegularFileProperty getOutputFile();
+
+        RegularFileProperty getCompileOutputFile();
+
+        Property<File> getRootDir();
+    }
+
+    private static final Pattern COMPILE_ERROR_FIRST_LINE_PATTERN =
+            Pattern.compile("^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: .*)$");
+    private static final Pattern COMPILE_ERROR_LAST_LINE_PATTERN = Pattern.compile("^\\d+ error(s)*$");
+    private static final Pattern COMPILE_ERROR_PATTERN =
+            Pattern.compile("(?m)^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: (.|\\s)*)");
+
+    private final ConcurrentMap<String, Boolean> startedCompileErrorsByTaskPath = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, StringBuilder> compilerErrorsByTaskPath = new ConcurrentHashMap<>();
+
+    public final <T extends AbstractCompile> void maybeCollectErrorMessage(T task, CharSequence charSequence) {
+        Matcher firstCompileErrorMatcher = COMPILE_ERROR_FIRST_LINE_PATTERN.matcher(charSequence);
+        if (firstCompileErrorMatcher.find()) {
+            String sourcePathFromError = firstCompileErrorMatcher.group("sourcePath");
+
+            // When running in parallel, javaCompileTasks will see the output of other tasks,
+            // see: https://github.com/gradle/gradle/issues/6068 for context.
+            // If the matched sourcePath is not part of the current javaCompileTask's sourceSet, we need to
+            // ignore it.
+            Set<String> taskCompiledSourcePaths = task.getSource().getFiles().stream()
+                    .map(File::getAbsolutePath)
+                    .collect(Collectors.toSet());
+            if (!taskCompiledSourcePaths.contains(sourcePathFromError)) {
+                return;
+            }
+            markCompileErrorStarted(task.getPath());
+            compilerErrorsByTaskPath
+                    .computeIfAbsent(task.getPath(), _k -> new StringBuilder())
+                    .append(charSequence);
+            return;
+        }
+        Matcher lastCompileErrorMessage = COMPILE_ERROR_LAST_LINE_PATTERN.matcher(charSequence);
+        if (isCompileErrorStarted(task.getPath()) && lastCompileErrorMessage.find()) {
+            markCompileErrorFinished(task.getPath());
+            return;
+        }
+        if (isCompileErrorStarted(task.getPath())) {
+            compilerErrorsByTaskPath
+                    .computeIfAbsent(task.getPath(), _k -> new StringBuilder())
+                    .append(charSequence);
+        }
```

Comment:
This part reads a little confusingly to me. When do we expect to not have seen the start of a compile error?

I think as written we could get rid of the repeated `isCompileErrorStarted(task.getPath())` check + return - perhaps something like:

```java
if (!isCompileErrorStarted(task.getPath())) {
    return;
}

if (lastCompileErrorMessage.find()) {
    // ...
    return;
}

// ...
```

Reply:
> When do we expect to not have seen the start of a compile error?
There could be some extra warnings as well added to the stdErr, will try to get an example but it makes sense, will refactor it.